https://github.com/kowainik/extensions/pull/80
From ae5376ffc9be056b9a0fd58f93cc61d924ae6239 Mon Sep 17 00:00:00 2001
From: hololeap <hololeap@users.noreply.github.com>
Date: Thu, 22 Sep 2022 12:56:45 -0600
Subject: [PATCH] Support Cabal-3.6

'hsSourceDirs' returns a list of 'SymbolicDir's in Cabal-3.6. This uses
CPP to check for a later version of Cabal and convert back to
'FilePath's if needed, using Distribution.Utils.Path.getSymbolicPath

Signed-off-by: hololeap <hololeap@users.noreply.github.com>
---
 extensions.cabal        | 2 +-
 src/Extensions/Cabal.hs | 8 ++++++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/extensions.cabal b/extensions.cabal
index db233bb..5adc9c9 100644
--- a/extensions.cabal
+++ b/extensions.cabal
@@ -69,7 +69,7 @@ library
                          Extensions.Types
 
   build-depends:       bytestring ^>= 0.10
-                     , Cabal >= 3.0 && < 3.5
+                     , Cabal >= 3.0 && < 3.7
                      , containers ^>= 0.6
                      , directory ^>= 1.3
                      , filepath ^>= 1.4
diff --git a/src/Extensions/Cabal.hs b/src/Extensions/Cabal.hs
index bd9d5c9..a031b66 100644
--- a/src/Extensions/Cabal.hs
+++ b/src/Extensions/Cabal.hs
@@ -53,6 +53,10 @@ import qualified Data.Map.Strict as Map
 import qualified Data.Text as Text
 import qualified Language.Haskell.Extension as Cabal
 
+#if MIN_VERSION_Cabal(3,6,0)
+import Distribution.Utils.Path (getSymbolicPath)
+#endif
+
 
 {- | Parse default extensions from a @.cabal@ file under given
 'FilePath'.
@@ -144,7 +148,11 @@ condTreeToExtensions
 condTreeToExtensions extractModules extractBuildInfo condTree = do
     let comp = condTreeData condTree
     let buildInfo = extractBuildInfo comp
+#if MIN_VERSION_Cabal(3,6,0)
+    let srcDirs = getSymbolicPath <$> hsSourceDirs buildInfo
+#else
     let srcDirs = hsSourceDirs buildInfo
+#endif
     let modules = extractModules comp ++
             map toModulePath (otherModules buildInfo ++ autogenModules buildInfo)
     let (safeExts, parsedExtensionsAll) = partitionEithers $ mapMaybe cabalToGhcExtension $ defaultExtensions buildInfo
